Examples supported sync connect families.

redboltz opened the pull request.

jonesmz commented on example/locked_cout.hpp:
```diff
@@ -0,0 +1,46 @@
+    std::unique_lock<std::mutex> lock_;
+    std::ostream& stream_;
+
+};
```
I've solved this problem this way in the past:

```cpp
struct locked
{
    std::ostream& cout() const { return std::cout; }
    std::ostream& cerr() const { return std::cerr; }

private:
    static std::mutex& mutex()
    {
        static std::mutex stdio_mutex;
        return stdio_mutex;
    }
    std::unique_lock<std::mutex> lock_{mutex()};
};

// Report a failure
void
fail(std::error_code ec)
{
    locked().cerr() << ec.message() << std::endl;
}
```

redboltz replied:
How does it work? I can't image client code of locked.

jonesmz replied:
Just like this:

```cpp
    locked().cerr() << ec.message() << std::endl;
```
You construct a new instance of it when you want to write, and call either cerr() or cout() to get the stream that you're writing to.

I'm not saying you should change your code to do this. I was simply showing you an example.

You might consider using the function-static variable for the mutex though, since it keeps the initialization of the mutex variable in one place.

jonesmz commented on example/locked_cout.hpp:
```diff
@@ -0,0 +1,46 @@
+         stream_(stream) {}
+
+    friend
+    locked_stream&& operator<<(locked_stream&& s, std::ostream& (*arg)(std::ostream&)) {
```
why the use of rvalue references here?

redboltz replied:
This is to write as follows:

```cpp
locked_cout() << data1 << data2;
```

locked_cout() is rvalue.

jonesmz commented:
I'm actually not understanding why multiple threads are needed here.

What's causing the deadlock when a single thread is used?

redboltz commented:
> What's causing the deadlock when a single thread is used?

Ok, let me explain.

The examples `*_both.cpp` contains both client and server. They use one `io_context ioc;`. At the end of the code, `ioc.run()` is called.

After I updated `async_connect()` and (sync)`connect()`, connect, tls handshake, and websocket handshake are processed synchronously. The old behavior is all of them are asynchronous.

And the example code uses `connect()`, not `async_connect()`. Because it is simpler.

If the client calls synchronous API such as `handshake`, then the code is blocked until the server finish server-side handshake processing. However, the client and server share the same `io_context`. So server had no chance to process server-side process.

So I introduced the new `io_context iocs;` and call `iocs.run()` from the server thread.
